File: test/test_particle.py
""" Implement tests for the Particle class. """

import pytest
import numpy as np

from nuchic.physics import Vector4, Vector3
from nuchic.physics import Particle, ParticleStatus
from nuchic.physics import PID, ParticleInfo
from nuchic._nuchic.utilities.constants import mN


def test_particle_init():
    """ Test particle initialization. """
    part = Particle(2212, Vector4(4, 3, 2, 1), Vector3(0, 0, 0), 0, [], [])
    assert part.pid() == PID(2212)
    assert part.momentum() == Vector4(4, 3, 2, 1)
    assert part.position() == Vector3(0, 0, 0)
    assert part.status() == ParticleStatus.background
    assert part.mothers() == []
    assert part.daughters() == []
    with pytest.raises(Exception):
        Particle(PID(2212), Vector4(4, 3, 2, 1), Vector3(0, 0, 0), None, None, None)


def test_particle_status():
    """ Test particle status codes. """
    part = Particle(pid=2212, momentum=Vector4(4, 3, 2, 1),
                    status=ParticleStatus.escaped)
    assert part.is_final()
    assert not part.is_background()
    assert not part.is_propagating()
    part.set_status(ParticleStatus.background)
    assert not part.is_final()
    assert part.is_background()
    assert not part.is_propagating()
    part.set_status(ParticleStatus.propagating)
    assert not part.is_final()
    assert not part.is_background()
    assert part.is_propagating()


def test_particle_mom():
    """ Test particle momentums. """
    part = Particle(2212, Vector4(3, 2, 1, 4))
    assert part.energy() == 4
    assert part.px() == 3
    assert part.py() == 2
    assert part.pz() == 1
    assert part.mass() == np.sqrt(2)
    assert part.beta() == Vector3(3.0/4.0, 0.5, 0.25)


def test_particle_prop():
    """ Test propagation of particles. """
    part = Particle(2212, Vector4(4, 3, 2, 1), Vector3(0, 0, 0), 0)
    part.propagate(0.1)
    part.back_propagate(0.1)
    assert part.radius() == 0


def test_particle_fzone():
    """ Test particle formation zone. """
    part = Particle(2212, Vector4(np.sqrt(10**2+mN**2), 0, 0, 10),
                    Vector3(0, 0, 0), 0)
    rap = 1
    p_t = 3
    m_t = np.sqrt(p_t**2 + mN**2)
    part_f = Particle(2212, Vector4(m_t*np.cosh(rap), p_t, 0, m_t*np.sinh(rap)),
                      Vector3(0, 0, 0), 0)
    assert not part.is_in_formation_zone()
    part.set_formation_zone(part.momentum(), part_f.momentum())
    f_zone = part.energy()/np.abs(mN**2 - part.momentum()*part_f.momentum())
    print(mN)
    assert part.formation_zone() == f_zone
    assert part.is_in_formation_zone()

File: test/test_particle.cc
// SPDX-FileCopyrightText: 2018-2026 Achilles Developers
// SPDX-License-Identifier: GPL-3.0-or-later

#include <sstream>

#include "catch2/catch_approx.hpp"
#include "catch2/catch_test_macros.hpp"
#include "catch2/matchers/catch_matchers.hpp"

#include "Achilles/Constants.hh"
#include "Achilles/Particle.hh"
#include "Approx.hh"

using achilles::FourVector;
using achilles::Particle;
using achilles::ThreePosition;
namespace units = achilles::units;
using namespace achilles::units::literals;

constexpr units::Energy energy = 1000_MeV;
constexpr units::Time timestep = 10000.0 * units::fm;

TEST_CASE("Formation Zone", "[Particle]") {
    Particle part{achilles::PID::proton(), {units::Energy{energy}, 100_MeV, 0_MeV, 0_MeV}};
    FourVector mom{energy, 0_MeV, 100_MeV, 0_MeV};

    part.SetFormationZone(part.Momentum(), mom);
    CHECK(part.InFormationZone());
    // Large enough number to ensure it exits the formation zone
    part.UpdateFormationZone(timestep);
    CHECK(!part.InFormationZone());
}

TEST_CASE("Properties", "[Particle]") {
    Particle part{achilles::PID::proton(),
                  {units::Energy{energy}, 100_MeV, 0_MeV, 0_MeV},
                  {0.0_fm, 1.0_fm, 0.0_fm}};

    SECTION("Momentum") {
        CHECK(part.Momentum() == FourVector{energy, 100_MeV, 0_MeV, 0_MeV});
        part.SetMomentum({energy, 0_MeV, 100_MeV, 0_MeV});
        CHECK(part.Momentum() == FourVector{energy, 0_MeV, 100_MeV, 0_MeV});
    }

    SECTION("Position") {
        CHECK(part.Position() == ThreePosition{0.0_fm, 1.0_fm, 0.0_fm});
        part.SetPosition({0.0_fm, 0.0_fm, 1.0_fm});
        CHECK(part.Position() == ThreePosition{0.0_fm, 0.0_fm, 1.0_fm});
    }

    SECTION("Status") {
        CHECK(part.Status() == achilles::ParticleStatus::background);
        CHECK(part.IsBackground());

        part.Status() = achilles::ParticleStatus::propagating;
        CHECK(part.IsPropagating());

        part.Status() = achilles::ParticleStatus::final_state;
        CHECK(part.IsFinal());
    }

    SECTION("History") {}
}

TEST_CASE("Propagate", "[Particle]") {
    Particle part{achilles::PID::proton(), {units::Energy{energy}, 100_MeV, 0_MeV, 0_MeV}};
    static constexpr double eps = 1e-8;

    SECTION("Time propagation") {
        part.Propagate(1.0 * units::fm);
        const double beta = (part.Momentum().P() / part.E()).native();
        CHECK_THAT(part.Position(),
                   IsVectorApprox<ThreePosition>(ThreePosition{beta * units::fm, 0.0_fm, 0.0_fm})
                       .margin(eps));
        part.BackPropagate(1.0 * units::fm);
        CHECK_THAT(
            part.Position(),
            IsVectorApprox<ThreePosition>(ThreePosition{0.0_fm, 0.0_fm, 0.0_fm}).margin(eps));
    }

    SECTION("Space propagation") {
        part.SpacePropagate(1.0 * units::fm);
        CHECK_THAT(
            part.Position(),
            IsVectorApprox<ThreePosition>(ThreePosition{1.0_fm, 0.0_fm, 0.0_fm}).margin(eps));
    }
}

TEST_CASE("I/O", "[Particle]") {
    Particle part{achilles::PID::proton(), {units::Energy{energy}, 100_MeV, 0_MeV, 0_MeV}};
    Particle part2;

    CHECK(part.ToString() == "Particle(2212, FourVector(1000.000000, 100.000000, 0.000000, "
                             "0.000000), ThreeVector(0.000000, 0.000000, 0.000000), 25)");

    std::stringstream ss;
    ss << part;
    ss >> part2;

    CHECK(part == part2);
}

TEST_CASE("Closest Approach", "[Particle]") {
    // A particle moving at beta = 1/hbarc in z, one fm-scale offset away.
    Particle part1(achilles::PID::proton(),
                   {1_MeV, 0_MeV, 0_MeV, units::Energy{1.0 / achilles::Constant::HBARC}});
    Particle part2(achilles::PID::proton(), {}, {3.0_fm, 2.0_fm, 1.0_fm});

    auto time = achilles::ClosestApproach(part1, part2);
    CHECK(time.in(units::fm) == Catch::Approx(achilles::Constant::HBARC));

    part1.Propagate(time);
    CHECK(part1.Position() == ThreePosition{0.0_fm, 0.0_fm, 1.0_fm});
}
